# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [29]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [30]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [31]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [32]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [33]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [34]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [35]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [36]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [37]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [38]:
data['children'].unique()


array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [39]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [40]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [41]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [42]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [43]:
data.duplicated().sum()

71

In [44]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [45]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [46]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [47]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [48]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [49]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

In [50]:


import plotly.express as px

def pivot(data, index):
    
    
    pivot = data.pivot_table(index = index, values = 'debt', aggfunc = ['count', 'sum', 'mean']).reset_index()
    pivot.columns = [index, 'Кол-во клиентов', 'Кол-во должников', 'Доля невозврата']
    display(pivot.sort_values('Доля невозврата', ascending = False))
    
    
    
    fig = px.bar(pivot, x = index, y = 'Доля невозврата', color = 'Доля невозврата', title = 'Доля невозврата по ' + index)
    fig.show()
    

pivot(data, 'children')

children  Кол-во клиентов  Кол-во должников  Доля невозврата
4         4               41                 4         0.097561
2         2             2052               194         0.094542
1         1             4808               444         0.092346
3         3              330                27         0.081818
0         0            14091              1063         0.075438
5         5                9                 0         0.000000

In [51]:
import plotly.express as px

def pivot(data, index):
    
    pivot = data.pivot_table(index = index, values = 'debt', aggfunc = ['count', 'sum', 'mean']).reset_index()
    pivot.columns = [index, 'Кол-во клиентов', 'Кол-во должников', 'Доля невозврата']
    display(pivot.sort_values('Доля невозврата', ascending = False))
    
    
    
    fig = px.bar(pivot, x = index, y = 'Доля невозврата', color = 'Доля невозврата', title = 'Доля невозврата по ' + index)
    fig.show()
    

pivot(data, 'family_status')
pivot(data, 'total_income_category')
pivot(data, 'purpose_category')

family_status  Кол-во клиентов  Кол-во должников  Доля невозврата
0  Не женат / не замужем             2796               273         0.097639
3       гражданский брак             4134               385         0.093130
4        женат / замужем            12261               927         0.075606
1              в разводе             1189                84         0.070648
2         вдовец / вдова              951                63         0.066246

total_income_category  Кол-во клиентов  Кол-во должников  Доля невозврата
4                     E               22                 2         0.090909
2                     C            15921              1353         0.084982
0                     A               25                 2         0.080000
1                     B             5014               354         0.070602
3                     D              349                21         0.060172

purpose_category  Кол-во клиентов  Кол-во должников  \
0    операции с автомобилем             4279               400   
2     получение образования             3988               369   
3        проведение свадьбы             2313               183   
1  операции с недвижимостью            10751               780   

   Доля невозврата  
0         0.093480  
2         0.092528  
3         0.079118  
1         0.072551

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [52]:
debt_children = pd.DataFrame()
debt_children['count_children'] = data.groupby('children')['debt'].count()
debt_children['sum_children'] = data.groupby('children')['debt'].sum()
debt_children['result'] = debt_children['sum_children'] / debt_children['count_children'] 
debt_children.sort_values('result', ascending = False)

count_children  sum_children    result
children                                        
4                     41             4  0.097561
2                   2052           194  0.094542
1                   4808           444  0.092346
3                    330            27  0.081818
0                  14091          1063  0.075438
5                      9             0  0.000000

**Вывод:** 

    Из проведенного анализа можно сделать следующие выводы:
    0) рассмотр вариант, когда детей в семье 5, не стоит учитывать в финальном результате, т.к выборка мала
    1) если у заемщика нет детей его шанс возврата кредита, выше чем если бы у него были дети в любом кол-ве
    2) наименьший шанс невозврата у заемщиков с семьей, если в этой семье 3 ребенка
    3) при 4 детях шанс невозврата наибольший, а именно 9.76%

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [53]:
#Для решения используется та же логика, отличие лишь в финальном представлении результатов
debt_data2 = data['debt'].groupby(data['family_status']).sum()
family_data = data['family_status'].groupby(data['family_status']).count()
print(debt_data2)

family_status
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
гражданский брак         385
женат / замужем          927
Name: debt, dtype: int64


In [54]:
debt_family_status = pd.DataFrame()
debt_family_status['sum_family_status'] = debt_data2 
debt_family_status['count_family_status'] = family_data
debt_family_status['result'] = ((debt_family_status['sum_family_status'] / debt_family_status['count_family_status'])*100).round(2) 
debt_family_status.sort_values('result', ascending = False)

sum_family_status  count_family_status  result
family_status                                                        
Не женат / не замужем                273                 2796    9.76
гражданский брак                     385                 4134    9.31
женат / замужем                      927                12261    7.56
в разводе                             84                 1189    7.06
вдовец / вдова                        63                  951    6.62

**Вывод:** 

    1) наименьший шанс невозврата долга у следующих групп:
    - вдова\вдовец - 6.62%
    - в разводе - 7.06%
    2) средний показатель невозврата у группы
    женат\замужем - 7.56%
    3) наибольший показатель невозврата долга у двух следующих групп
    - гражданский брак - 9.31%
    - не женат\не замужем - 9.76%
    1.Можно сказать о следующей зависимости - люди не бывшие в официальном браке чаще имеют проблемы с невозвратом
    2.В то время как люди которые уже были в браке, но по какой-то причине сейчас в нем не состоят имеют наименьшие проблемы с возвратом кредита

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Старый вариант:

In [55]:
#Эксперементально разбивал на группы по разному доходу и оптимальный вариат разбитие по 50000.
#Разбитие на группы реализую с помощью функции
#Дальнейшие шаги идентичны предидущим группам
def total_income_fin(value):
    value = value // 50000
    return value

data['total_income_id'] = data['total_income'].apply(total_income_fin)
debt_total_income = pd.DataFrame()
debt_total_income['sum'] = data.groupby('total_income_id')['debt'].sum()
debt_total_income['count'] = data.groupby('total_income_id')['debt'].count()
debt_total_income['result'] = debt_total_income['sum'] / debt_total_income['count']
debt_total_income.sort_values('result', ascending = False)


sum  count    result
total_income_id                      
44                 1      1  1.000000
20                 1      4  0.250000
16                 1      6  0.166667
17                 1      6  0.166667
12                 3     21  0.142857
2                619   7110  0.087060
3                403   4738  0.085057
14                 1     12  0.083333
6                 51    617  0.082658
1                331   4073  0.081267
10                 5     68  0.073529
7                 24    329  0.072948
4                162   2242  0.072257
8                 13    195  0.066667
5                 88   1323  0.066515
0                 23    371  0.061995
9                  4    111  0.036036
11                 1     45  0.022222
15                 0      6  0.000000
13                 0     27  0.000000
18                 0      3  0.000000
19                 0      3  0.000000
21                 0      5  0.000000
22                 0      2  0.000000
23                 0      1  0.000000
24                 0      2  0.000000
25                 0      2  0.000000
27                 0      1  0.000000
28                 0      1  0.000000
31                 0      2  0.000000
34                 0      3  0.000000
45                 0      1  0.000000

Новый варинт

In [56]:
debt_total_income = pd.DataFrame()
debt_total_income['sum'] = data.groupby('total_income_category')['debt'].sum()
debt_total_income['count'] = data.groupby('total_income_category')['debt'].count()
debt_total_income['result'] = debt_total_income['sum'] / debt_total_income['count']
debt_total_income.sort_values('result', ascending = False)

sum  count    result
total_income_category                       
E                         2     22  0.090909
C                      1353  15921  0.084982
A                         2     25  0.080000
B                       354   5014  0.070602
D                        21    349  0.060172

**Вывод:** 

Зависимости нет, вероятности всех групп близки друг к другу

Не совсем понял, что подразумевается под категориями B и C (у меня не было групп по алфавиту), тем не менее воспользовался qcut, чтобы разбить на равные группы и получил красивый результат.

Очень классная вещь, спасибо за рекомендацию. 

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [57]:
debt_purpose_category = pd.DataFrame()
debt_purpose_category['sum_purpose'] = data.groupby('purpose_category')['debt'].sum()
debt_purpose_category['count_purpose'] = data.groupby('purpose_category')['debt'].count()
debt_purpose_category['result'] = ((debt_purpose_category['sum_purpose'] / debt_purpose_category['count_purpose'])*100).round(2) 
debt_purpose_category.sort_values('result', ascending = False)

sum_purpose  count_purpose  result
purpose_category                                            
операции с автомобилем            400           4279    9.35
получение образования             369           3988    9.25
проведение свадьбы                183           2313    7.91
операции с недвижимостью          780          10751    7.26

**Вывод:** 

операции связанные с авто и получение образования чаще получают невозврат, тогда как кредит на недвижимость и проведение свадьбы имеют больший процент возврата кредита в срок

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

    Пропускив данных могут обуславливаться следующими причинами:
    1) Ошибка работы системы
    2) Конфиденциальность данных (госслуажие и прочие)
    3) невозмодность их обработки 
    4) невнимательность персонала
    5) человеческий фактор
    По своей сути факторов повлиявших на появление пропусков может быть очень много

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

При расчете медианы используются весь набор данных,а также медиана слабовосприимчива к выбросам, что сильно помогает при работе с большими данными. Если бы в данных заданиях мы высчитывали бы среднее значение, оно было бы не репрезентативным из-за очень малых или очень больших значений, которые исказили бы результат 

### Шаг 4: общий вывод.

    Новый вывод:
    В ходе выполненной работы были найдены следующие важные зависиомти:
        1. Cемейный статус влияет на вероятность платежей по кредиту в срок - люди, бывшие в браке, но в данный момент по какой-либо причине сейчас в нем не состоящие чаще плятят в срок чем те, кто в данный момент находится в браке. Заемщики которые никогда небыли в браке платят хуже всего.  
        2. Бездетные заемщики чаще платят по кредиту в срок
        3. Операции связанные с авто и получение образования чаще получают невозврат, тогда как кредит на недвижимость и проведение свадьбы имеют больший процент возврата кредита в срок